lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -1145,6 +1105,93 @@ bool AssociatedTypeInference::checkConstrainedExtension(ExtensionDecl *ext) {
   llvm_unreachable("unhandled result");
 }
 
+AssociatedTypeDecl *AssociatedTypeInference::completeSolution(
+    ArrayRef<AssociatedTypeDecl *> unresolvedAssocTypes, unsigned reqDepth) {
+  // Examine the solution for errors and attempt to compute abstract type
+  // witnesses for associated types that are still lacking an entry.
+  llvm::SmallVector<AbstractTypeWitness, 2> abstractTypeWitnesses;
+  for (auto *const assocType : unresolvedAssocTypes) {
+    const auto typeWitness = typeWitnesses.begin(assocType);
+    if (typeWitness != typeWitnesses.end()) {
+      // The solution contains an error.
+      if (typeWitness->first->hasError()) {
+        return assocType;
+      }
+
+      continue;
+    }
+
+    // Try to compute the type without the aid of a specific potential witness.
+    if (const auto typeWitness = computeAbstractTypeWitness(assocType)) {
+      abstractTypeWitnesses.push_back(typeWitness.getValue());
+
+      // Record the type witness immediately to make it available
+      // for substitutions into other tentative type witnesses.
+      typeWitnesses.insert(assocType, {typeWitness->type, reqDepth});
+      continue;
+    }
+
+    // The solution is incomplete.
+    return assocType;
+  }
+
+  const auto typeSubstFn = [&](SubstitutableType *type) {
+    if (type->isEqual(proto->getSelfInterfaceType()))
+      return adoptee;
+
+    return Type();
+  };
+
+  // Custom LookupConformanceFn to avoid re-looking up the conformance
+  // we're checking.
+  const auto lookupConfFn = [&](CanType dependentType,
+                                Type conformingReplacementType,
+                                ProtocolDecl *conformedProtocol) {
+    if (conformedProtocol == proto &&
```

Comment:
Is this really necessary? Conformance lookup doesn't trigger computation of anything, it just returns an existing conformance even if its type witnesses have not yet been filled in.

Reply:
This is somewhat a paranoia I have about conformance lookup determinism dating back to [this](https://github.com/apple/swift/pull/30470#issuecomment-601476004) comment by Joe :)